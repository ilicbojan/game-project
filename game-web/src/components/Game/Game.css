.game-results {
  display: flex;
  flex-direction: row;
  gap: 2rem;
}

@media screen and (max-width: 768px) {
  .game-results {
    flex-direction: column;
  }
}
