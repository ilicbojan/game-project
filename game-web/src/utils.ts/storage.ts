export function saveToStorage<T>(key: string, value: T) {
  localStorage.setItem(key, JSON.stringify(value));
}

export function loadFromStorage<T>(key: string, fallback: T): T {
  const data = localStorage.getItem(key);
  if (!data) return fallback;

  return JSON.parse(data, (_key, value) => {
    // Convert ISO strings to Date
    if (
      typeof value === "string" &&
      /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/.test(value)
    ) {
      const date = new Date(value);
      if (!isNaN(date.getTime())) return date;
    }
    return value;
  });
}
