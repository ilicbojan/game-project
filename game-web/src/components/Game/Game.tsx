import { useEffect, useState } from "react";
import GameApi from "../../api/gameApi";
import {
  ChoiceItem,
  GameResult,
  PlayRequest,
  PlayResponse,
  Score,
} from "../../types/game";
import { loadFromStorage, saveToStorage } from "../../utils.ts/storage";
import Choices from "../Choices/Choices";
import Result from "../Result/Result";
import Scoreboard from "../Scoreboard/Scoreboard";
import History from "../History/History";
import Info from "../Info/Info";
import "./Game.css";
import { toast } from "react-toastify";

const Game = () => {
  const [choices, setChoices] = useState<ChoiceItem[]>([]);
  const [result, setResult] = useState<PlayResponse | null>(null);
  const [loadingChoices, setLoadingChoices] = useState(false);
  const [loadingResult, setLoadingResult] = useState(false);
  const [score, setScore] = useState<Score>(
    loadFromStorage("score", { win: 0, lose: 0, tie: 0 })
  );
  const [history, setHistory] = useState<GameResult[]>(
    loadFromStorage("history", [])
  );

  useEffect(() => {
    setLoadingChoices(true);
    GameApi.getChoices()
      .then(setChoices)
      .catch(() => {})
      .finally(() => setLoadingChoices(false));
  }, []);

  const handlePlay = async (choiceId: number) => {
    setLoadingResult(true);
    setResult(null);

    try {
      const request: PlayRequest = { player: choiceId };
      const response = await GameApi.play(request);
      setResult(response);

      const playerName =
        choices.find((c) => c.id === response.player)?.name || "";
      const computerName =
        choices.find((c) => c.id === response.computer)?.name || "";

      const updatedScore = { ...score };
      updatedScore[response.result as keyof Score] += 1;
      setScore(updatedScore);
      saveToStorage("score", updatedScore);

      const newHistory: GameResult = {
        player: playerName,
        computer: computerName,
        result: response.result,
        date: new Date(),
      };

      // Remove the oldest entry if there is 10 already
      let trimmedHistory = history;
      if (history.length >= 10) {
        trimmedHistory = history.slice(0, 9);
      }
      const updatedHistory = [newHistory, ...trimmedHistory];
      setHistory(updatedHistory);
      saveToStorage("history", updatedHistory);
    } catch (error) {
    } finally {
      setLoadingResult(false);
    }
  };

  const handleRandomChoice = async () => {
    setLoadingResult(true);
    setResult(null);

    try {
      const randomChoice = await GameApi.getRandomChoice();
      if (randomChoice) {
        await handlePlay(randomChoice.id);
      }
    } catch (error) {
    } finally {
      setLoadingResult(false);
    }
  };

  const handleReset = () => {
    setHistory([]);
    setScore({ win: 0, lose: 0, tie: 0 });
    setResult(null);
    saveToStorage("score", { win: 0, lose: 0, tie: 0 });
    saveToStorage("history", []);
  };

  return (
    <div className="game">
      <Info handleReset={handleReset} />
      <Choices
        choices={choices}
        onChoiceClick={handlePlay}
        onRandomClick={handleRandomChoice}
        loadingChoices={loadingChoices}
        loadingResult={loadingResult}
      />
      <div className="game-results">
        <Result result={result} choices={choices} loading={loadingResult} />
        <Scoreboard score={score} />
      </div>
      <History history={history} />
    </div>
  );
};

export default Game;
