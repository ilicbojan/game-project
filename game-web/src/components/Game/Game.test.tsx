import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import Game from "./Game";
import GameApi from "../../api/gameApi";
import * as storage from "../../utils.ts/storage";

// Mock components to isolate logic
jest.mock("../Choices/Choices", () => ({
  __esModule: true,
  default: (props: { onChoiceClick: (id: number) => void }) => (
    <div data-testid="choices" onClick={() => props.onChoiceClick(1)}>
      Choices
    </div>
  ),
}));
jest.mock("../Result/Result", () => ({
  __esModule: true,
  default: () => <div data-testid="result">Result</div>,
}));
jest.mock("../Scoreboard/Scoreboard", () => ({
  __esModule: true,
  default: () => <div data-testid="scoreboard">Scoreboard</div>,
}));
jest.mock("../History/History", () => ({
  __esModule: true,
  default: () => <div data-testid="history">History</div>,
}));

// Mock API
jest.mock("../../api/gameApi");

const choicesMock = [
  { id: 1, name: "rock" },
  { id: 2, name: "paper" },
];

const playResponseMock = {
  player: 1,
  computer: 2,
  result: "win",
};

describe("Game component", () => {
  beforeEach(() => {
    (GameApi.getChoices as jest.Mock).mockResolvedValue(choicesMock);
    (GameApi.play as jest.Mock).mockResolvedValue(playResponseMock);
    jest
      .spyOn(storage, "loadFromStorage")
      .mockImplementation((key, fallback) => {
        if (key === "score") return { win: 0, lose: 0, tie: 0 };
        if (key === "history") return [];
        return fallback;
      });
    jest.spyOn(storage, "saveToStorage").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Game />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with all components", async () => {
    render(<Game />);

    expect(await screen.findByTestId("choices")).toBeInTheDocument();
    expect(screen.getByTestId("result")).toBeInTheDocument();
    expect(screen.getByTestId("scoreboard")).toBeInTheDocument();
    expect(screen.getByTestId("history")).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /reset game/i })
    ).toBeInTheDocument();
  });

  it("calls GameApi.play and updates score and history", async () => {
    render(<Game />);

    const choices = await screen.findByTestId("choices");

    await fireEvent.click(choices);

    await waitFor(() => {
      expect(GameApi.play).toHaveBeenCalledWith({ player: 1 });
    });

    expect(storage.saveToStorage).toHaveBeenCalledWith("score", {
      win: 1,
      lose: 0,
      tie: 0,
    });

    expect(storage.saveToStorage).toHaveBeenCalledWith("history", [
      expect.objectContaining({
        date: expect.any(Date),
        player: "rock",
        computer: "paper",
        result: "win",
      }),
    ]);
  });

  it("resets the game when clicking reset", async () => {
    render(<Game />);

    act(() => {
      fireEvent.click(screen.getByText(/reset game/i));
    });

    expect(storage.saveToStorage).toHaveBeenCalledWith("score", {
      win: 0,
      lose: 0,
      tie: 0,
    });

    expect(storage.saveToStorage).toHaveBeenCalledWith("history", []);
  });
});
