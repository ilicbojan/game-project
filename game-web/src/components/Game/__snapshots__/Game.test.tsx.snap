// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="game"
  >
    <div
      class="info section"
    >
      <div
        class="info-header"
      >
        <h3>
          Game Info
        </h3>
        <div
          class="info-buttons"
        >
          <button
            aria-controls="info-content"
            aria-expanded="false"
            class="info-accordion-toggle"
          >
            Show More
          </button>
          <button
            class="info-reset-button"
          >
            Reset Game ðŸ”ƒ
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="choices"
    >
      Choices
    </div>
    <div
      class="game-results"
    >
      <div
        data-testid="result"
      >
        Result
      </div>
      <div
        data-testid="scoreboard"
      >
        Scoreboard
      </div>
    </div>
    <div
      data-testid="history"
    >
      History
    </div>
  </div>
</DocumentFragment>
`;
